File: app/src/main/java/com/degree/application/view/fragment/CheckFragment.java
package com.degree.application.view.fragment;


import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import com.degree.application.R;
import com.degree.application.service.ContractContext;
import com.degree.application.service.ContractValidatorManager;
import com.degree.application.service.ContractValidatorManager.ValidatorListener;
import com.degree.application.view.MainActivity;

import io.rmiri.buttonloading.ButtonLoading;

import static android.widget.Toast.LENGTH_LONG;
import static com.degree.application.utils.LoadingDialogUtils.displayLoadingDialog;
import static com.degree.application.utils.LoadingDialogUtils.hideLoadingDialog;


public class CheckFragment extends Fragment implements ValidatorListener {

    private ContractValidatorManager contractValidatorManager;

    private EditText verifyDegreeId, studentName,
            birthday, graduationDate, degreeLabel;
    private String contractAddress;

    private ButtonLoading verify;

    public CheckFragment() {
        contractValidatorManager = ContractValidatorManager.getInstance();
    }

    /**
     * Returns a new instance of this fragment for the given section
     * number.
     */
    public static CheckFragment newInstance() {
        return new CheckFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_check, container, false);

        verifyDegreeId = rootView.findViewById(R.id.degreeId);
        studentName = rootView.findViewById(R.id.studentName);
        birthday = rootView.findViewById(R.id.birthday);
        graduationDate = rootView.findViewById(R.id.graduationDate);
        degreeLabel = rootView.findViewById(R.id.degreeLabel);

        verify = rootView.findViewById(R.id.verify);
        verify.setOnButtonLoadingListener(new ButtonLoading.OnButtonLoadingListener() {
            @Override
            public void onClick() {
                displayLoadingDialog(getActivity());
                contractValidatorManager.verify(
                        CheckFragment.this,
                        verifyDegreeId.getText().toString(),
                        studentName.getText().toString(),
                        birthday.getText().toString(),
                        graduationDate.getText().toString(),
                        degreeLabel.getText().toString()
                );
            }

            @Override
            public void onStart() {

            }

            @Override
            public void onFinish() {

            }
        });
        return rootView;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            verifyDegreeId.setText(ContractContext.getDegreeId());
            studentName.setText(ContractContext.getStudentName());
            birthday.setText(ContractContext.getBirthday());
            graduationDate.setText(ContractContext.getGraduationDate());
            degreeLabel.setText(ContractContext.getDegreeLabel());

        }
    }

    @Override
    public void onDegreeValid() {
        nextPage(getString(R.string.button_ok));
    }

    @Override
    public void onDegreeInvalid() {
        nextPage(getString(R.string.degree_nok));
    }

    @Override
    public void onError(String message) {
        getActivity().runOnUiThread(() -> {
            verify.setProgress(false);
            hideLoadingDialog();
            displayMessage("Error during validation " + message);
        });
    }

    private void displayMessage(String message) {
        Toast.makeText(getContext(), message, LENGTH_LONG).show();
    }

    private void nextPage(String message) {
        getActivity().runOnUiThread(() -> {
            hideLoadingDialog();
            ((MainActivity) getActivity()).nextCurrentItem();
            displayMessage(message);
            verify.setProgress(false);
        });
    }
}

File: app/src/main/java/io/smartpal/smartdegree/view/fragment/CheckFragment.kt
package io.smartpal.smartdegree.view.fragment

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.smartpal.smartdegree.R
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.AppCompatEditText
import android.widget.Toast
import io.smartpal.smartdegree.contract.SmartDegreeService
import io.smartpal.smartdegree.contract.SmartDegreeModel
import io.smartpal.smartdegree.contract.SmartDegreeListener
import io.smartpal.smartdegree.helpers.ContractLoader
import io.smartpal.smartdegree.helpers.LoadingDialog.displayLoadingDialog
import io.smartpal.smartdegree.helpers.LoadingDialog.hideLoadingDialog
import io.smartpal.smartdegree.view.activity.MainActivity
import io.smartpal.smartdegree.view.model.ContractViewModel


class CheckFragment : Fragment() {


    private lateinit var viewModel: ContractViewModel

    private lateinit var degreeId: AppCompatEditText
    private lateinit var studentName: AppCompatEditText
    private lateinit var birthday: AppCompatEditText
    private lateinit var graduationDate: AppCompatEditText
    private lateinit var degreeLabel: AppCompatEditText
    private lateinit var contractAddress: AppCompatEditText

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_check, container, false)
        setupViewModel()
        setupForm(rootView)
        return rootView
    }

    private fun setupViewModel() {
        viewModel = ViewModelProviders.of(activity!!).get(ContractViewModel::class.java)
    }

    private fun setupForm(view: View) {
        with(view) {
            degreeId = findViewById(R.id.degreeId)
            studentName = findViewById(R.id.studentName)
            birthday = findViewById(R.id.birthday)
            graduationDate = findViewById(R.id.graduationDate)
            degreeLabel = findViewById(R.id.degreeLabel)
            contractAddress = findViewById(R.id.contractAddress)
        }

        viewModel.getContractModel().observe(this, Observer { model: ContractLoader.ContractModel? ->
            var modelSmartDegree = model as SmartDegreeModel
            degreeId.setText(modelSmartDegree.degreeId)
            studentName.setText(modelSmartDegree.studentName)
            birthday.setText(modelSmartDegree.birthday)
            graduationDate.setText(modelSmartDegree.graduationDate)
            degreeLabel.setText(modelSmartDegree.degreeLabel)
            contractAddress.setText(modelSmartDegree.contractAddress)
        })

    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            activity?.findViewById<FloatingActionButton>(R.id.fab)?.let {
                it.visibility = View.VISIBLE
                it.display
                it.setOnClickListener {
                    context?.let {
                        displayLoadingDialog(it)
                        callContract(it)
                    }
                }
            }
        }
    }

    private fun callContract(context: Context) {
        var model = SmartDegreeModel(
                degreeId.text.toString(),
                studentName.text.toString(),
                birthday.text.toString(),
                graduationDate.text.toString(),
                degreeLabel.text.toString(),
                contractAddress.text.toString()
        )

        SmartDegreeService.verify(context,
                model, object : SmartDegreeListener {
            override fun onResponseReceived(isValidDegree: Boolean) {
                viewModel.getCallContractResult().postValue(isValidDegree)
                nextPage()
            }

            override fun onError(message: String?) {
                if (message != null) {
                    showError(message)
                }
            }
        })
    }

    private fun nextPage() {
        activity?.runOnUiThread {
            hideLoadingDialog()
            (activity as MainActivity).nextItem()
        }
    }

    private fun showError(message: String) {
        activity?.runOnUiThread {
            hideLoadingDialog()
            Toast.makeText(context, message, Toast.LENGTH_LONG).show()
        }
    }

    companion object {
        /**
         * Returns a new instance of this fragment for the given section
         * number.
         */
        fun newInstance(): CheckFragment {
            return CheckFragment()
        }
    }

}
